lib/elixir_sense/core/metadata_builder.ex

```diff
@@ -1663,6 +1668,98 @@ defmodule ElixirSense.Core.MetadataBuilder do
     {ast, {vars, match_context}}
   end
 
+  def infer_type_from_guards(guard_ast, vars, state) do
+    type_info = type_information_from_guards(guard_ast, state)
+
+    Enum.reduce(type_info, vars, fn {var, type}, acc ->
+      index = Enum.find_index(acc, &(&1.name == var))
+
+      if index,
+        do: List.update_at(acc, index, &Map.put(&1, :type, type)),
+        else: acc
+    end)
+  end
+
+  # A guard expression can be in either these form:
+  #        :and                          :or
+  #      /     \            or          /   \              or      guard_expr
+  # guard_expr  guard_expr        guard_expr  guard_expr
+  #
+  # type information from :and subtrees are mergeable
+  # type information from :or subtrees are discarded
+  defp type_information_from_guards({:and, _, [guard_l, guard_r]}, state) do
+    left = type_information_from_guards(guard_l, state)
+    right = type_information_from_guards(guard_r, state)
+
+    Keyword.merge(left, right, fn _k, v1, v2 ->
+      case {v1, v2} do
+        # func my_func(x) when is_map_key(x, :a) and is_map_key(x, :b)
+        {{:map, fields1, _}, {:map, fields2, _}} ->
+          {:map, Enum.uniq_by(fields1 ++ fields2, &elem(&1, 0)), nil}
+
+        # In case we can't merge, just pick one
+        _ ->
+          v1
+      end
+    end)
+  end
+
+  defp type_information_from_guards({:or, _, [_guard_l, _guard_r]}, _state), do: []
```

Comment:
Let's use union type here
```suggestion
  defp type_information_from_guards({:or, _, [guard_l, guard_r]}, _state), do: [{:union, [guard_l, _guard_r]}]
```

Reply:
Fixed in 11428b98127e2ad9b554149115d95ef161c77ab1

Reply:
Should I move [this logic](https://github.com/elixir-lsp/elixir_sense/commit/11428b98127e2ad9b554149115d95ef161c77ab1#diff-b689b035238e6e8efc041c52f31533f0b7534ce543b3cb0179f8039bf1bf2302R1707-R1719) into the Binding module as well

Reply:
> Should I move [this logic](https://github.com/elixir-lsp/elixir_sense/commit/11428b98127e2ad9b554149115d95ef161c77ab1#diff-b689b035238e6e8efc041c52f31533f0b7534ce543b3cb0179f8039bf1bf2302R1707-R1719) into the Binding module as well

That could be a good idea

Reply:
I guess I'll keep it as is for now, the binding logic is pretty complicated already